src/sync/coll.rs

```diff
@@ -0,0 +1,316 @@
+///
+/// `Collection` uses [`std::sync::Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html) internally,
+/// so it can safely be shared across threads. For example:
+///
+/// ```rust
+/// # use bson::{bson, doc};
+/// # use mongodb::{sync::Client, error::Result};
+/// #
+/// # fn start_workers() -> Result<()> {
+/// # let client = Client::with_uri_str("mongodb://example.com")?;
+/// let coll = client.database("items").collection("in_stock");
+///
+/// for i in 0..5 {
+///     let coll_ref = coll.clone();
+///
+///     std::thread::spawn(move || {
+///         // Perform operations with `coll_ref`. For example:
+///         coll_ref.insert_one(doc! { "x": i }, None);
+///     });
+/// }
+/// #
+/// # // Technically we should join the threads here, but for the purpose of the example, we'll just
+/// # // sleep for a bit.
+/// # std::thread::sleep(std::time::Duration::from_secs(3));
+/// # Ok(())
+/// # }
+/// ```
+
+#[derive(Clone, Debug)]
+pub struct Collection {
+    async_collection: AsyncCollection,
+}
+
+impl Collection {
+    pub(crate) fn new(async_collection: AsyncCollection) -> Self {
+        Self { async_collection }
+    }
+
+    /// Gets the name of the `Collection`.
+    pub fn name(&self) -> &str {
+        self.async_collection.name()
+    }
+
+    /// Gets the namespace of the `Collection`.
+    ///
+    /// The namespace of a MongoDB collection is the concatenation of the name of the database
+    /// containing it, the '.' character, and the name of the collection itself. For example, if a
+    /// collection named "bar" is created in a database named "foo", the namespace of the collection
+    /// is "foo.bar".
+    pub fn namespace(&self) -> Namespace {
+        self.async_collection.namespace()
+    }
+
+    /// Gets the selection criteria of the `Collection`.
+    pub fn selection_criteria(&self) -> Option<&SelectionCriteria> {
+        self.async_collection.selection_criteria()
+    }
+
+    /// Gets the read concern of the `Collection`.
+    pub fn read_concern(&self) -> Option<&ReadConcern> {
+        self.async_collection.read_concern()
+    }
+
+    /// Gets the write concern of the `Collection`.
+    pub fn write_concern(&self) -> Option<&WriteConcern> {
+        self.async_collection.write_concern()
+    }
+
+    /// Drops the collection, deleting all data, users, and indexes stored in it.
+    pub fn drop(&self, options: impl Into<Option<DropCollectionOptions>>) -> Result<()> {
+        RUNTIME.block_on(self.async_collection.drop(options.into()))
+    }
+
+    /// Runs an aggregation operation.
+    ///
+    /// See the documentation [here](https://docs.mongodb.com/manual/aggregation/) for more
+    /// information on aggregations.
+    pub fn aggregate(
+        &self,
+        pipeline: impl IntoIterator<Item = Document>,
+        options: impl Into<Option<AggregateOptions>>,
+    ) -> Result<Cursor> {
+        let pipeline: Vec<Document> = pipeline.into_iter().collect();
```

Comment:
I think we can just pass the pipeline in directly, right? The async collection's `aggregate` method also accepts `impl IntoIterator<Item = Document>`.

Reply:
In order to pass it directly, we'd need to add the `Send` requirement to the `impl trait` clause in the sync API, which I figured we didn't want to do.

relevant compiler message:
```
error: future cannot be sent between threads safely
   --> src/sync/coll.rs:121:23
    |
116 |         pipeline: impl IntoIterator<Item = Document>,
    |                   ---------------------------------- help: consider further restricting this bound: `impl IntoIterator<Item = Document> + std::marker::Send`
...
121 |             .block_on(self.async_collection.aggregate(pipeline, options.into()))
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ future returned by `aggregate` is not `Send`
    |
    = help: within `impl core::future::future::Future`, the trait `std::marker::Send` is not implemented for `impl IntoIterator<Item = Document>`
note: future is not `Send` as this value is used across an await
```

Reply:
Fair enough. I'd be surprised if people had containers of documents that were not `Send`, but like, if a user wants to use `cursor.map(|doc| doc.unwrap()` as their aggregation pipeline, I'm not going to stop them.